### PR TITLE
[WIP] GstGLPlayer: Gstreamer OpenGL video provider

### PR DESCRIPTION
**Pull request by hunternet93**
This PR contains GstGLPlayer, a new Video provider for Kivy that uses Gstreamer's integrated OpenGL support to upload frames into Kivy's OpenGL context without hogging the main thread.

Needs some work, see #5336.

### CODE REVIEW

**Comment by tito** [comment]
Did it work for you? I'm stuck in the _on_video_frame called in update_texture on the first load. Look like a deadlock. I'm using the `examples/widgets/videoplayer.py` example.

**Comment by hunternet93** [comment]
That's yet another bug, forgot to mention it. Using the `play` property to start it causes the hang, but setting `play` to True after creating the widget works. No idea why.

**Comment by tito** [comment]
Cause you have a deadlock with the GIL. Look at both backtrace (i stripped out the one that are not interesting, check at the end).

We have 2 threads, the main Python thread, and one of the gstreamer thread (14). The gstreamer thread at some point call you on_appsink_preroll, and then go back to Python, then still in the stack, the Python call _get_duration in gstreamer. And my guts tell me because you have sync message emission enabled, i think gstreamer can't then call anything back on Python, even if the GIL got released ?

```
Thread 14 (Thread 0x7f8ed11d5700 (LWP 20480)):
#0  0x00007f8f09578029 in syscall () at /usr/lib/libc.so.6
#1  0x00007f8f01eda22c in  () at /usr/lib/libglib-2.0.so.0
#2  0x00007f8eff5853d1 in  () at /usr/lib/libgstbase-1.0.so.0
#3  0x00007f8eff585ccf in  () at /usr/lib/libgstbase-1.0.so.0
#4  0x00007f8effca3854 in  () at /usr/lib/libgstreamer-1.0.so.0
#5  0x00007f8effca4257 in gst_pad_set_active () at /usr/lib/libgstreamer-1.0.so.0
#6  0x00007f8effc81f1f in  () at /usr/lib/libgstreamer-1.0.so.0
#7  0x00007f8effc9489c in gst_iterator_fold () at /usr/lib/libgstreamer-1.0.so.0
#8  0x00007f8effc82a79 in  () at /usr/lib/libgstreamer-1.0.so.0
#9  0x00007f8effc84847 in  () at /usr/lib/libgstreamer-1.0.so.0
#10 0x00007f8effc84b06 in  () at /usr/lib/libgstreamer-1.0.so.0
#11 0x00007f8eff5846e6 in  () at /usr/lib/libgstbase-1.0.so.0
#12 0x00007f8effc86ade in gst_element_change_state () at /usr/lib/libgstreamer-1.0.so.0
#13 0x00007f8effc8723f in  () at /usr/lib/libgstreamer-1.0.so.0
#14 0x00007f8effc646d8 in  () at /usr/lib/libgstreamer-1.0.so.0
#15 0x00007f8effc86ade in gst_element_change_state () at /usr/lib/libgstreamer-1.0.so.0
#16 0x00007f8effc8723f in  () at /usr/lib/libgstreamer-1.0.so.0
#17 0x00007f8effc646d8 in  () at /usr/lib/libgstreamer-1.0.so.0
#18 0x00007f8effc86ade in gst_element_change_state () at /usr/lib/libgstreamer-1.0.so.0
#19 0x00007f8effc8723f in  () at /usr/lib/libgstreamer-1.0.so.0
#20 0x00007f8effc646d8 in  () at /usr/lib/libgstreamer-1.0.so.0
#21 0x00007f8ef5ca9668 in  () at /usr/lib/gstreamer-1.0/libgstplayback.so
#22 0x00007f8effc86ade in gst_element_change_state () at /usr/lib/libgstreamer-1.0.so.0
#23 0x00007f8effc8723f in  () at /usr/lib/libgstreamer-1.0.so.0
#24 0x00007f8effc646d8 in  () at /usr/lib/libgstreamer-1.0.so.0
#25 0x00007f8ef5c9a6c9 in  () at /usr/lib/gstreamer-1.0/libgstplayback.so
#26 0x00007f8effc86ade in gst_element_change_state () at /usr/lib/libgstreamer-1.0.so.0
#27 0x00007f8effc8723f in  () at /usr/lib/libgstreamer-1.0.so.0
#28 0x00007f8effc646d8 in  () at /usr/lib/libgstreamer-1.0.so.0
#29 0x00007f8effc86ade in gst_element_change_state () at /usr/lib/libgstreamer-1.0.so.0
#30 0x00007f8effc8723f in  () at /usr/lib/libgstreamer-1.0.so.0
#31 0x00007f8efff69092 in __pyx_f_4kivy_3lib_11gstglplayer_12_gstglplayer_11GstGLPlayer__get_duration (__pyx_v_self=0x7f8ef6ce4470) at kivy/lib/gstglplayer/_gstglplayer.c:5427
#32 0x00007f8efff6a0fb in __pyx_pf_4kivy_3lib_11gstglplayer_12_gstglplayer_11GstGLPlayer_18get_duration (__pyx_v_self=0x7f8ef6ce4470) at kivy/lib/gstglplayer/_gstglplayer.c:4970
#33 0x00007f8efff6a0fb in __pyx_pw_4kivy_3lib_11gstglplayer_12_gstglplayer_11GstGLPlayer_19get_duration (__pyx_v_self=0x7f8ef6ce4470, unused=<optimized out>)
    at kivy/lib/gstglplayer/_gstglplayer.c:4934
#34 0x00007f8f0a268f9a in call_function (oparg=<optimized out>, pp_stack=0x7f8ed11d2f88) at Python/ceval.c:4336
#35 0x00007f8f0a268f9a in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#36 0x00007f8f0a26a345 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7f8ed11d30f8, func=0x7f8ef6a867d0) at Python/ceval.c:4437
#37 0x00007f8f0a26a345 in call_function (oparg=<optimized out>, pp_stack=0x7f8ed11d30f8) at Python/ceval.c:4372
#38 0x00007f8f0a26a345 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#39 0x00007f8f0a26ac3e in PyEval_EvalCodeEx (co=0x7f8f005c2c30, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=1, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3584
#40 0x00007f8f0a1e5a61 in function_call (func=0x7f8f005c7320, arg=0x7f8ef6754b10, kw=0x0) at Objects/funcobject.c:523
#41 0x00007f8f0a1b5e93 in PyObject_Call (func=0x7f8f005c7320, arg=<optimized out>, kw=<optimized out>) at Objects/abstract.c:2547
#42 0x00007f8f0a1b915c in call_function_tail (args=0x7f8ef6754b10, callable=0x7f8f005c7320) at Objects/abstract.c:2579
---Type <return> to continue, or q <return> to quit---
#43 0x00007f8f0a1b915c in PyObject_CallFunction (callable=0x7f8f005c7320, format=<optimized out>) at Objects/abstract.c:2603
#44 0x00007f8f0a20137b in _PyObject_GenericGetAttrWithDict (obj=0x7f8ef6a68610, name=0x7f8f054ed660, dict=0x0) at Objects/object.c:1439
#45 0x00007f8f0a2651b5 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2543
#46 0x00007f8f0a26ac3e in PyEval_EvalCodeEx (co=0x7f8f005b83b0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=2, kws=0x7f8ef66f73a8, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3584
#47 0x00007f8f0a26a1f7 in fast_function (nk=<optimized out>, na=2, n=<optimized out>, pp_stack=0x7f8ed11d3758, func=0x7f8ef6a86d70) at Python/ceval.c:4447
#48 0x00007f8f0a26a1f7 in call_function (oparg=<optimized out>, pp_stack=0x7f8ed11d3758) at Python/ceval.c:4372
#49 0x00007f8f0a26a1f7 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#50 0x00007f8f0a26ac3e in PyEval_EvalCodeEx (co=0x7f8f005b84b0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=2, kws=0x7f8f0a703068, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3584
#51 0x00007f8f0a1e5b68 in function_call (func=0x7f8ef6a86e60, arg=0x7f8ef66ec3f8, kw=0x7f8ef66f2910) at Objects/funcobject.c:523
#52 0x00007f8f0a1b5e93 in PyObject_Call (func=0x7f8ef6a86e60, arg=<optimized out>, kw=<optimized out>) at Objects/abstract.c:2547
#53 0x00007f8f0a1c864f in instancemethod_call (func=0x7f8ef6a86e60, arg=0x7f8ef66ec3f8, kw=0x7f8ef66f2910) at Objects/classobject.c:2602
#54 0x00007f8f058cd060 in __Pyx_PyObject_Call (func=0x7f8efbaaec30, arg=<optimized out>, kw=<optimized out>) at kivy/_event.c:16410
#55 0x00007f8f058d399f in __pyx_f_4kivy_6_event_14EventObservers__dispatch (__pyx_v_self=__pyx_v_self@entry=0x7f8ef66f0208, __pyx_v_f=__pyx_v_f@entry=0x7f8efbaaec30, __pyx_v_slargs=__pyx_v_slargs@entry=0x7f8f0a51f270 <_Py_NoneStruct>, __pyx_v_skwargs=__pyx_v_skwargs@entry=0x7f8f0a51f270 <_Py_NoneStruct>, __pyx_v_obj=__pyx_v_obj@entry=0x7f8ef6a68610, __pyx_v_value=__pyx_v_value@entry=0x7f8f0a51f270 <_Py_NoneStruct>, __pyx_v_largs=0x7f8f0a703050, __pyx_v_kwargs=0x7f8ef66f2910) at kivy/_event.c:13564
#56 0x00007f8f058d8e0b in __pyx_f_4kivy_6_event_14EventObservers_dispatch (__pyx_v_self=__pyx_v_self@entry=0x7f8ef66f0208, __pyx_v_obj=__pyx_v_obj@entry=0x7f8ef6a68610, __pyx_v_value=__pyx_v_value@entry=0x7f8f0a51f270 <_Py_NoneStruct>, __pyx_v_largs=__pyx_v_largs@entry=0x7f8f0a703050, __pyx_v_kwargs=__pyx_v_kwargs@entry=0x7f8ef66f2910, __pyx_v_stop_on_true=__pyx_v_stop_on_true@entry=1) at kivy/_event.c:14035
#57 0x00007f8f058e09c0 in __pyx_pf_4kivy_6_event_15EventDispatcher_24dispatch (__pyx_v_kwargs=0x7f8ef66f2910, __pyx_v_largs=0x7f8f0a703050, __pyx_v_event_type=<optimized out>, __pyx_v_self=0x7f8ef6a68610) at kivy/_event.c:8146
#58 0x00007f8f058e09c0 in __pyx_pw_4kivy_6_event_15EventDispatcher_25dispatch (__pyx_v_self=0x7f8ef6a68610, __pyx_args=<optimized out>, __pyx_kwds=<optimized out>)
    at kivy/_event.c:8099
#59 0x00007f8f0a2691e5 in call_function (oparg=<optimized out>, pp_stack=0x7f8ed11d3e98) at Python/ceval.c:4352
#60 0x00007f8f0a2691e5 in PyEval_EvalFrameEx (f=<optimized out>, f@entry=0x7f8ecc00f4a0, throwflag=<optimized out>, throwflag@entry=0) at Python/ceval.c:2989
#61 0x00007f8efff6aa13 in __Pyx_PyFunction_FastCallNoKw (globals=<optimized out>, na=<optimized out>, args=<optimized out>, co=<optimized out>)
    at kivy/lib/gstglplayer/_gstglplayer.c:6655
#62 0x00007f8efff6aa13 in __Pyx_PyFunction_FastCallDict (func=func@entry=0x7f8ef6a86938, args=<optimized out>, nargs=<optimized out>, kwargs=0x0)
    at kivy/lib/gstglplayer/_gstglplayer.c:6695
#63 0x00007f8efff6c69c in __pyx_f_4kivy_3lib_11gstglplayer_12_gstglplayer_11GstGLPlayer_process_sample (__pyx_v_self=0x7f8ef6ce4470, __pyx_v_sample=<optimized out>)
    at kivy/lib/gstglplayer/_gstglplayer.c:4134
#64 0x00007f8efff6a3fe in __pyx_f_4kivy_3lib_11gstglplayer_12_gstglplayer__on_appsink_preroll (__pyx_v_appsink=<optimized out>, __pyx_v_user_data=0x7f8ef6ce4470)
    at kivy/lib/gstglplayer/_gstglplayer.c:1643
#65 0x00007f8eff7d011a in  () at /usr/lib/libgstapp-1.0.so.0
#66 0x00007f8eff58bac9 in gst_base_sink_do_preroll () at /usr/lib/libgstbase-1.0.so.0
#67 0x00007f8eff58c4e5 in  () at /usr/lib/libgstbase-1.0.so.0
#68 0x00007f8eff58dca5 in  () at /usr/lib/libgstbase-1.0.so.0
#69 0x00007f8eff58edc9 in  () at /usr/lib/libgstbase-1.0.so.0
#70 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#71 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#72 0x00007f8eff599013 in  () at /usr/lib/libgstbase-1.0.so.0
#73 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#74 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#75 0x00007f8effc8dac5 in gst_proxy_pad_chain_default () at /usr/lib/libgstreamer-1.0.so.0
#76 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#77 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#78 0x00007f8effc8dac5 in gst_proxy_pad_chain_default () at /usr/lib/libgstreamer-1.0.so.0
#79 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#80 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#81 0x00007f8eff599013 in  () at /usr/lib/libgstbase-1.0.so.0
#82 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#83 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#84 0x00007f8eff599013 in  () at /usr/lib/libgstbase-1.0.so.0
#85 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#86 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#87 0x00007f8eff599013 in  () at /usr/lib/libgstbase-1.0.so.0
---Type <return> to continue, or q <return> to quit---
#88 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#89 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#90 0x00007f8eff599013 in  () at /usr/lib/libgstbase-1.0.so.0
#91 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#92 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#93 0x00007f8effc8dac5 in gst_proxy_pad_chain_default () at /usr/lib/libgstreamer-1.0.so.0
#94 0x00007f8effc9f6ab in  () at /usr/lib/libgstreamer-1.0.so.0
#95 0x00007f8effca7be3 in gst_pad_push () at /usr/lib/libgstreamer-1.0.so.0
#96 0x00007f8ef444cbc5 in  () at /usr/lib/gstreamer-1.0/libgstcoreelements.so
#97 0x00007f8effcd4083 in  () at /usr/lib/libgstreamer-1.0.so.0
#98 0x00007f8f01ebd4b0 in  () at /usr/lib/libglib-2.0.so.0
#99 0x00007f8f01ebcae5 in  () at /usr/lib/libglib-2.0.so.0
#100 0x00007f8f09f55049 in start_thread () at /usr/lib/libpthread.so.0
#101 0x00007f8f0957cf0f in clone () at /usr/lib/libc.so.6

Thread 1 (Thread 0x7f8f0a744440 (LWP 20334)):
#0  0x00007f8f0a2181a0 in PyType_IsSubtype (a=0x7f8f0a513bc0 <_PyExc_IndexError>, b=0x7f8f0a513bc0 <_PyExc_IndexError>) at Objects/typeobject.c:1180
#1  0x00007f8f0a219f32 in type___subclasscheck__ (type=<optimized out>, inst=<optimized out>) at Objects/typeobject.c:629
#2  0x00007f8f0a1b5e93 in PyObject_Call (func=0x7f8ef6746f80, arg=<optimized out>, kw=<optimized out>) at Objects/abstract.c:2547
#3  0x00007f8f0a1baa66 in PyObject_CallFunctionObjArgs (callable=0x7f8ef6746f80) at Objects/abstract.c:2774
#4  0x00007f8f0a1babbb in PyObject_IsSubclass (derived=0x7f8f0a513bc0 <_PyExc_IndexError>, cls=0x7f8f0a513bc0 <_PyExc_IndexError>) at Objects/abstract.c:3055
#5  0x00007f8f0a277a18 in PyErr_GivenExceptionMatches (err=0x7f8f0a513bc0 <_PyExc_IndexError>, exc=0x7f8f0a513bc0 <_PyExc_IndexError>) at Python/errors.c:119
#6  0x00007f8f0a267afd in cmp_outcome (w=0x7f8f0a513bc0 <_PyExc_IndexError>, v=0x7f8f0a513bc0 <_PyExc_IndexError>, op=<optimized out>) at Python/ceval.c:4855
#7  0x00007f8f0a267afd in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2576
#8  0x00007f8f0a26ac3e in PyEval_EvalCodeEx (co=0x7f8efc1e0930, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=1, kws=0x7f8ef66f3208, kwcount=1, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3584
#9  0x00007f8f0a26a1f7 in fast_function (nk=<optimized out>, na=1, n=<optimized out>, pp_stack=0x7fffffffd2e8, func=0x7f8efbaac488) at Python/ceval.c:4447
#10 0x00007f8f0a26a1f7 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffd2e8) at Python/ceval.c:4372
#11 0x00007f8f0a26a1f7 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#12 0x00007f8f0a26a345 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffd458, func=0x7f8f05502140) at Python/ceval.c:4437
---Type <return> to continue, or q <return> to quit---
#13 0x00007f8f0a26a345 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffd458) at Python/ceval.c:4372
#14 0x00007f8f0a26a345 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#15 0x00007f8f0a26a345 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffd5c8, func=0x7f8f055021b8) at Python/ceval.c:4437
#16 0x00007f8f0a26a345 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffd5c8) at Python/ceval.c:4372
#17 0x00007f8f0a26a345 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#18 0x00007f8f0a26a345 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffd738, func=0x7f8efbab12a8) at Python/ceval.c:4437
#19 0x00007f8f0a26a345 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffd738) at Python/ceval.c:4372
#20 0x00007f8f0a26a345 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#21 0x00007f8f0a26a345 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffd8a8, func=0x7f8efbab1410) at Python/ceval.c:4437
#22 0x00007f8f0a26a345 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffd8a8) at Python/ceval.c:4372
#23 0x00007f8f0a26a345 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#24 0x00007f8f0a26ac3e in PyEval_EvalCodeEx (co=0x7f8f06c2ddb0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=0, kws=0x7f8efc1e9d70, kwcount=0, defs=0x7f8f05629068, defcount=2, closure=0x0) at Python/ceval.c:3584
#25 0x00007f8f0a26a1f7 in fast_function (nk=<optimized out>, na=0, n=<optimized out>, pp_stack=0x7fffffffdac8, func=0x7f8f055025f0) at Python/ceval.c:4447
#26 0x00007f8f0a26a1f7 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffdac8) at Python/ceval.c:4372
#27 0x00007f8f0a26a1f7 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#28 0x00007f8f0a26a345 in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffdc38, func=0x7f8f00e4a050) at Python/ceval.c:4437
#29 0x00007f8f0a26a345 in call_function (oparg=<optimized out>, pp_stack=0x7fffffffdc38) at Python/ceval.c:4372
#30 0x00007f8f0a26a345 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at Python/ceval.c:2989
#31 0x00007f8f0a26ac3e in PyEval_EvalCodeEx (co=0x7f8f0a66ceb0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=0, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3584
#32 0x00007f8f0a26ad52 in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at Python/ceval.c:669
#33 0x00007f8f0a28b450 in run_mod (arena=0x6a3fb0, flags=0x7fffffffdf30, locals=0x7f8f0a6d8168, globals=0x7f8f0a6d8168, filename=<optimized out>, mod=0x6811e8)
    at Python/pythonrun.c:1376
#34 0x00007f8f0a28b450 in PyRun_FileExFlags (fp=0x63db50, filename=<optimized out>, start=<optimized out>, globals=0x7f8f0a6d8168, locals=0x7f8f0a6d8168, closeit=1, flags=0x7fffffffdf30) at Python/pythonrun.c:1362
#35 0x00007f8f0a28b62f in PyRun_SimpleFileExFlags (fp=0x63db50, filename=0x7fffffffe40c "examples/widgets/videoplayer.py", closeit=1, flags=0x7fffffffdf30)
    at Python/pythonrun.c:948
#36 0x00007f8f0a2a0fd4 in Py_Main (argc=<optimized out>, argv=<optimized out>) at Modules/main.c:645
#37 0x00007f8f094af4ca in __libc_start_main () at /usr/lib/libc.so.6
#38 0x0000000000400729 in _start ()
```

**Comment by tito** [comment]
If i move the on_load call within _update, it continues:

```
    def _update(self, dt):
        if self._next_texture:
            self._texture = self._next_texture
            self._next_texture = None

            # There's probably a better way to do this, TODO
            if not self._on_load_called:
                self.dispatch('on_load')
                self._on_load_called = True

            self.dispatch('on_frame')
```

I guess GST don't like to dispatch a message while dispatching a message.

**Comment by tito** [comment]
Do you have a crash when the video ends by itself?

```
Thread 14 (Thread 0x7f7e8e462700 (LWP 2982)):
#0  0x00007f7ebd278d25 in __Pyx_PyFunction_FastCallDict (func=func@entry=0x7f7eb3d94398, args=args@entry=0x7f7e8e460160, nargs=nargs@entry=1, kwargs=0x0)
    at kivy/lib/gstglplayer/_gstglplayer.c:6720
#1  0x00007f7ebd27a692 in __Pyx_PyObject_CallOneArg (arg=<optimized out>, func=0x7f7eb3d94398) at kivy/lib/gstglplayer/_gstglplayer.c:6848
#2  0x00007f7ebd27a692 in __pyx_f_4kivy_3lib_11gstglplayer_12_gstglplayer__on_appsink_eos (__pyx_v_appsink=<optimized out>, __pyx_v_user_data=0x7f7eb3ff5470)
    at kivy/lib/gstglplayer/_gstglplayer.c:1550
#3  0x00007f7ebcade683 in  () at /usr/lib/libgstapp-1.0.so.0
#4  0x00007f7ebc88fc14 in  () at /usr/lib/libgstbase-1.0.so.0
#5  0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#6  0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#7  0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#8  0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#9  0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#10 0x00007f7ebc8abd92 in  () at /usr/lib/libgstbase-1.0.so.0
#11 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#12 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#13 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#14 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#15 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#16 0x00007f7ebcfb71c8 in  () at /usr/lib/libgstreamer-1.0.so.0
#17 0x00007f7ebcfb2896 in gst_pad_forward () at /usr/lib/libgstreamer-1.0.so.0
#18 0x00007f7ebcfb299e in gst_pad_event_default () at /usr/lib/libgstreamer-1.0.so.0
#19 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#20 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#21 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#22 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#23 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#24 0x00007f7ebcfb71c8 in  () at /usr/lib/libgstreamer-1.0.so.0
#25 0x00007f7ebcfb2896 in gst_pad_forward () at /usr/lib/libgstreamer-1.0.so.0
#26 0x00007f7ebcfb299e in gst_pad_event_default () at /usr/lib/libgstreamer-1.0.so.0
#27 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#28 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#29 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#30 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#31 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#32 0x00007f7ebc8abd92 in  () at /usr/lib/libgstbase-1.0.so.0
---Type <return> to continue, or q <return> to quit---
#33 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#34 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#35 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#36 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#37 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#38 0x00007f7ebc8abd92 in  () at /usr/lib/libgstbase-1.0.so.0
#39 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#40 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#41 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#42 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#43 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#44 0x00007f7ebc8abd92 in  () at /usr/lib/libgstbase-1.0.so.0
#45 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#46 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#47 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#48 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#49 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#50 0x00007f7ebc8abd92 in  () at /usr/lib/libgstbase-1.0.so.0
#51 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#52 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#53 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#54 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#55 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#56 0x00007f7ebcfb71c8 in  () at /usr/lib/libgstreamer-1.0.so.0
#57 0x00007f7ebcfb2896 in gst_pad_forward () at /usr/lib/libgstreamer-1.0.so.0
#58 0x00007f7ebcfb299e in gst_pad_event_default () at /usr/lib/libgstreamer-1.0.so.0
#59 0x00007f7eb2fca068 in  () at /usr/lib/gstreamer-1.0/libgstplayback.so
#60 0x00007f7ebcfac407 in  () at /usr/lib/libgstreamer-1.0.so.0
#61 0x00007f7ebcfac881 in  () at /usr/lib/libgstreamer-1.0.so.0
#62 0x00007f7ebcfacc0a in  () at /usr/lib/libgstreamer-1.0.so.0
#63 0x00007f7ebcfaa820 in  () at /usr/lib/libgstreamer-1.0.so.0
#64 0x00007f7ebcfb6c90 in gst_pad_push_event () at /usr/lib/libgstreamer-1.0.so.0
#65 0x00007f7eb175add3 in  () at /usr/lib/gstreamer-1.0/libgstcoreelements.so
#66 0x00007f7ebcfe2083 in  () at /usr/lib/libgstreamer-1.0.so.0
#67 0x00007f7ebf1cc4b0 in  () at /usr/lib/libglib-2.0.so.0
#68 0x00007f7ebf1cbae5 in  () at /usr/lib/libglib-2.0.so.0
#69 0x00007f7ec7264049 in start_thread () at /usr/lib/libpthread.so.0
#70 0x00007f7ec688bf0f in clone () at /usr/lib/libc.so.6

```

**Comment by hunternet93** [comment]
Yep, haven't figured that one out yet. On my machine it uses massive amounts of CPU until I `kill -9` it.

**Comment by tito** [comment]
Ok. Then try to edit the issue and make a list of all the things that need to be fixed/sorted-of :)

File: kivy/core/window/_window_sdl2.pyx [hunk 1]
```diff
@@ -28,6 +29,9 @@ cdef class _WindowSDL2Storage:
         self.surface = NULL
         self.win_flags = 0
         self.event_filter = None
+    
+    def get_sdl2_window_pointer(self):
```
[the remark this answers is not in the document]
Comment (Cheaterman) [reply kivy/core/window/_window_sdl2.pyx]: It is used in kivy.lib.gstglplayer in _gstglplayer.pyx :-)

**Comment by Cheaterman** [comment]
Sooo. I spent most of my day working on this, and there are two main issues:
1°) A massive memory leak for a yet unknown reason (I thought it was because you keep a reference to all Texture instances you create, but removing this isn't sufficient)
2°) Garbage video output - I can see the images are indeed coming from my video, but they look terribly wrong, I suspect it is because of a wrong color format conversion somewhere or something like that

**Comment by hunternet93** [comment]
My code probably doesn't unref / free objects created by Gstreamer, as a Python programmer I'm used to letting the interpreter worry about that sort of thing.

Not sure about #2, Gstreamer's usually pretty good about getting colorspace conversion right and the caps are set to always convert to RGBA.

I'll see if I have time to fix the things @tito mentioned over the weekend, but I'd appreciate all the help I can get on the C / Cython front.

**Comment by hunternet93** [comment]
Closing due to age

**Pull request closed without merging**